src/protocol/ipa/mod.rs

```diff
@@ -760,158 +762,141 @@ pub mod tests {
-                let random_breakdown_key = if is_trigger_report {
-                    0
-                } else {
-                    rng.gen_range(0..MAX_BREAKDOWN_KEY)
-                };
-                records_for_user.push(TestRawDataRecord {
-                    user_id: random_user_id,
-                    timestamp: random_timestamp,
-                    is_trigger_report,
-                    breakdown_key: random_breakdown_key,
-                });
-            }
-
-            // sort in reverse time order
-            records_for_user.sort_unstable_by(|a, b| b.timestamp.cmp(&a.timestamp));
+        let random_seed = thread_rng().gen();
+        println!("Using random seed: {random_seed}");
 
-            let mut encountered_trigger = false;
-            let mut contributed_once = false;
-            for record in &records_for_user {
-                if contributed_once {
-                    continue;
+        for per_user_cap in [1, 3] {
+            println!("Running with a per-user-cap of {per_user_cap}.");
+
+            let mut rng = StdRng::seed_from_u64(random_seed);
+
+            let mut raw_data = Vec::with_capacity(NUM_USERS * MAX_RECORDS_PER_USER);
+            let mut expected_results = vec![0_u32; MAX_BREAKDOWN_KEY];
+
+            for _ in 0..NUM_USERS {
+                let random_user_id = rng.gen_range(0..MAX_USER_ID);
+                let num_records_for_user = rng.gen_range(1..MAX_RECORDS_PER_USER);
+                let mut records_for_user = Vec::with_capacity(num_records_for_user);
+                for _ in 0..num_records_for_user {
+                    let random_timestamp = rng.gen_range(0..SECONDS_IN_EPOCH);
+                    let is_trigger_report = rng.gen::<bool>();
+                    let random_breakdown_key = if is_trigger_report {
+                        0
+                    } else {
+                        rng.gen_range(0..MAX_BREAKDOWN_KEY)
+                    };
+                    let trigger_value = if is_trigger_report {
+                        rng.gen_range(1..MAX_TRIGGER_VALUE)
+                    } else {
+                        0
+                    };
+                    records_for_user.push(TestRawDataRecord {
+                        user_id: random_user_id,
+                        timestamp: random_timestamp,
+                        is_trigger_report,
+                        breakdown_key: random_breakdown_key,
+                        trigger_value,
+                    });
                 }
 
-                if record.is_trigger_report {
-                    encountered_trigger = true;
-                } else if encountered_trigger {
-                    expected_results[record.breakdown_key] += 1;
-                    contributed_once = true;
-                    encountered_trigger = false;
+                // sort in reverse time order
+                records_for_user.sort_unstable_by(|a, b| b.timestamp.cmp(&a.timestamp));
```

Comment:
This (from the `for _ in 0..NUM_USERS` onward) looks like something that could be its own function.  Maybe `random_user_records(...)`.  I'll leave it to you as to whether you want to pass constants as arguments or move them to the function.